src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,238 @@
+/-
+Copyright (c) 2021 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Ashvni Narayanan
+-/
+
+import number_theory.bernoulli
+
+/-!
+# Bernoulli polynomials
+
+The Bernoulli polynomials are an important tool obtained from Bernoulli numbers.
+
+## Mathematical overview
+
+The $n^{th}$ Bernoulli polynomial is defined as
+$$ B_n(X) = ∑_{k = 0}^n {n χoose k} (-1)^k * B_k * X^{n - k} $$
+where $B_k$ is the $k^{th}$ Bernoulli number. The Bernoulli polynomials are generating functions,
+$$ t * e^{tX} / (e^t - 1) = ∑_{n = 0}^{∞} B_n(X) * ¼{t^n}{n!} $$
+
+## Implementation detail
+
+The Bernoulli (negative) numbers are first defined, `bernoulli_neg`, in order to make proving
+theorems using Bernoulli polynomials easier.
+
+## Main theorems
+
+`sum_bernoulli_neg : ∑ k in range n, (n.choose k : ℚ) * bernoulli_neg k = 0`
+`sum_bernoulli_poly : ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k = `
+  `polynomial.monomial n (n + 1 : ℚ)`
+`exp_bernoulli_poly : power_series.mk (λ n, (polynomial.aeval t ((1 / nat.factorial n : ℚ) • `
+`  bernoulli_poly n))) * (exp A - 1) = X * rescale t (exp A) `
+-/
+
+noncomputable theory
+open_locale big_operators
+open_locale nat
+
+namespace finset
+
+lemma dependent_double_sum {M : Type*} [add_comm_monoid M]
```

Comment:
Perhaps `sum_Ico_Ico_comm` would be a more discoverable name. A pictorial representation might help too:
```
    j
    a     b       a     b
i a [=====)     a ⎴⎴⎴⎴⎴⎴ 
     [====)       ⏝|||||
      [===)        ⏝||||
       [==)  =      ⏝|||
        [=)          ⏝||
         [)           ⏝|
  b       |     b      ⏝
```

Reply:
Sorry I don't understand, do you want me to put a pictorial representation in the docstring? I have changed the name

Reply:
Yeah, I was suggesting it as a docstring, but only if you think it actually helps

Reply:
I think I'd be tempted to just document it as saying that the two ways of summing over (i,j) in the range a<=i<=j<b are equal.

Reply:
It would be neat if we could state both as equal to  `∑ i j (hai : a ≤ i) (hij : i ≤ j) (hij : j < b)`, but that's a long way from being legal at the moment.

Reply:
Done!